File: projects/Ляшук/3lab/app/src/main/java/com/example/a3lab/ListFragment.java
package com.example.a3lab;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.example.a3lab.databinding.FragmentListBinding;
import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ListFragment extends Fragment {

    private FragmentListBinding binding;
    private PostAdapter adapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentListBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        adapter = new PostAdapter(new ArrayList<>(), post -> {
            Bundle bundle = new Bundle();
            bundle.putSerializable("post", post);
            Navigation.findNavController(view).navigate(R.id.action_listFragment_to_detailFragment, bundle);
        });

        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.recyclerView.setAdapter(adapter);

        binding.buttonLoad.setOnClickListener(v -> loadData());
    }

    private void loadData() {
        String baseUrl = binding.editTextUrl.getText().toString();
        if (!baseUrl.endsWith("/")) {
            baseUrl += "/";
        }

        String endpoint = binding.editTextEndpoint.getText().toString();
        String limitStr = binding.editTextLimit.getText().toString();
        Integer limit = null;
        if (!limitStr.isEmpty()) {
            limit = Integer.parseInt(limitStr);
        }

        Retrofit retrofit;
        try {
            retrofit = new Retrofit.Builder()
                    .baseUrl(baseUrl)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        } catch (Exception e) {
            Toast.makeText(getContext(), "Invalid Base URL", Toast.LENGTH_SHORT).show();
            return;
        }

        ApiService apiService = retrofit.create(ApiService.class);
        Call<JsonElement> call = apiService.getRawData(endpoint, limit);

        call.enqueue(new Callback<JsonElement>() {
            @Override
            public void onResponse(Call<JsonElement> call, Response<JsonElement> response) {
                if (response.isSuccessful() && response.body() != null) {
                    JsonElement json = response.body();
                    List<Post> dataList = new ArrayList<>();
                    Gson gson = new Gson();

                    if (json.isJsonArray()) {
                        // Формат JSONPlaceholder: [...]
                        Type listType = new TypeToken<List<Post>>(){}.getType();
                        dataList = gson.fromJson(json, listType);
                    } else if (json.isJsonObject()) {
                        // Формат Rick and Morty: { "results": [...] }
                        ApiResponse apiResponse = gson.fromJson(json, ApiResponse.class);
                        if (apiResponse != null && apiResponse.getResults() != null) {
                            dataList = apiResponse.getResults();
                        }
                    }

                    if (!dataList.isEmpty()) {
                        adapter.setPosts(dataList);
                    } else {
                        Toast.makeText(getContext(), "No data found in expected fields", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(getContext(), "Error: " + response.code(), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<JsonElement> call, Throwable t) {
                Toast.makeText(getContext(), "Failure: " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: projects/Гончеренок/lab3/src_/JsonListApp/app/src/main/java/com/example/jsonlistapp/ui/ListFragment.kt
package com.example.jsonlistapp.ui

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.jsonlistapp.databinding.FragmentListBinding
import com.example.jsonlistapp.model.Post
import com.example.jsonlistapp.network.RetrofitClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ListFragment : Fragment() {

    interface OnPostClickListener {
        fun onPostClick(postId: Int, userId: Int, title: String, body: String)
    }

    private var _binding: FragmentListBinding? = null
    private val binding get() = _binding!!

    private val postList = mutableListOf<Post>()
    private lateinit var adapter: PostAdapter
    private var listener: OnPostClickListener? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnPostClickListener) {
            listener = context
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        adapter = PostAdapter(postList) {
            listener?.onPostClick(it.id, it.userId, it.title, it.body)
        }

        binding.recyclerViewPosts.layoutManager = LinearLayoutManager(requireContext())
        binding.recyclerViewPosts.adapter = adapter

        binding.btnLoadData.setOnClickListener {
            loadPosts()
        }

        binding.btnClear.setOnClickListener {
            postList.clear()
            adapter.notifyDataSetChanged()
        }
    }

    private fun loadPosts() {
        binding.progressBar.visibility = View.VISIBLE

        RetrofitClient.apiService.getPosts().enqueue(object : Callback<List<Post>> {
            override fun onResponse(call: Call<List<Post>>, response: Response<List<Post>>) {
                binding.progressBar.visibility = View.GONE

                if (response.isSuccessful && response.body() != null) {
                    postList.clear()
                    postList.addAll(response.body()!!)
                    adapter.notifyDataSetChanged()
                    Toast.makeText(requireContext(), "Данные загружены", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(requireContext(), "Ошибка загрузки", Toast.LENGTH_LONG).show()
                }
            }

            override fun onFailure(call: Call<List<Post>>, t: Throwable) {
                binding.progressBar.visibility = View.GONE
                Toast.makeText(requireContext(), "Ошибка сети", Toast.LENGTH_LONG).show()
            }
        })
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}